test_rmw_implementation/test/test_init_shutdown.cpp

```diff
@@ -28,20 +28,94 @@
 # define CLASSNAME(NAME, SUFFIX) NAME
 #endif
 
-class CLASSNAME (TestInitShutdown, RMW_IMPLEMENTATION) : public ::testing::Test {};
+class CLASSNAME (TestInitShutdown, RMW_IMPLEMENTATION) : public ::testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    options = rmw_get_zero_initialized_init_options();
+    rmw_ret_t ret = rmw_init_options_init(&options, rcutils_get_default_allocator());
+    ASSERT_EQ(RMW_RET_OK, ret) << rcutils_get_error_string().str;
+  }
+
+  void TearDown() override
+  {
+    rmw_ret_t ret = rmw_init_options_fini(&options);
+    EXPECT_EQ(RMW_RET_OK, ret) << rcutils_get_error_string().str;
+  }
+
+  rmw_init_options_t options;
+};
+
+TEST_F(CLASSNAME(TestInitShutdown, RMW_IMPLEMENTATION), init_with_bad_arguments) {
+  rmw_context_t context = rmw_get_zero_initialized_context();
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_init(nullptr, &context));
+  rcutils_reset_error();
+
+  context = rmw_get_zero_initialized_context();
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_init(&options, nullptr));
+  rcutils_reset_error();
+
+  const char * implementation_identifier = options.implementation_identifier;
+  options.implementation_identifier = "not-a-real-rmw-implementation-identifier";
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, rmw_init(&options, &context));
+  options.implementation_identifier = implementation_identifier;
+  rcutils_reset_error();
+}
+
+TEST_F(CLASSNAME(TestInitShutdown, RMW_IMPLEMENTATION), shutdown_with_bad_arguments) {
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_shutdown(nullptr));
+  rcutils_reset_error();
 
-TEST_F(CLASSNAME(TestInitShutdown, RMW_IMPLEMENTATION), init_shutdown) {
   rmw_context_t context = rmw_get_zero_initialized_context();
-  rmw_init_options_t options = rmw_get_zero_initialized_init_options();
-  rmw_ret_t ret = rmw_init_options_init(&options, rcutils_get_default_allocator());
+  EXPECT_NE(RMW_RET_OK, rmw_shutdown(&context));
```

Comment:
Is there a specific error you are expecting? RMW_RET_ERROR?

Reply:
Based on [`rmw_shutdown()` API contract](https://github.com/ros2/rmw/blob/b05f7eacd0e3e4209ffdcac40f3d353e932511b3/rmw/include/rmw/init.h#L86-L106), an RMW implementation could equally return `RMW_RET_INVALID_ARGUMENT` or `RMW_RET_INCORRECT_IMPLEMENTATION`. That's why I don't check for it.

Reply:
Since you already test the incorrect_implementation error below, wouldn't it make more sense to test the invalid_argument error here? Set context.implementation_identifier to a known valid identifier, then check that it returns invalid argument?

Reply:
Hmm, problem is that we get into bad initialization territory. Here I deal with the four (4) known context states: zero-initialized, initialized and not shutdown, initialized and shutdown, finalized. Anything outside that is UB. So if we want to be specific about the assertion we have to further clarify API contracts i.e. what it means for a context to be invalid.

Reply:
If you don't want to monkey around with the identifier field to force the invalid argument, then I recommend to at least narrow down the expect statement to only two errors you expect. Something like:

```
EXPECT_TRUE((ret == RMW_RET_INVALID_ARGUMENT || 
             ret == RMW_RET_INCORRECT_IMPLEMENTATION))
```

Reply:
Sure, we can do that. See  https://github.com/ros2/rmw_implementation/pull/107/commits/9bff7f27014246338bf1fe84322993a574db9f3a